### ISSUE TITLE
Split `Poll` into `Poll` and `Register`.

### ISSUE THREAD

**Issue by carllerche** [tokio-rs/mio#757]
As part of [rewriting the readiness queue](https://github.com/carllerche/mio/commit/43a69f218957ea7f43bb821734dbfbad09078c6b), `Poll::poll` acquired a lock. This is because `Poll` needed to be `Sync` so that I/O resources could be merged from other threads.

However, `Poll` really provides two separate capabilities:

* Registering I/O resources.
* Polling for I/O events.

Only I/O registration needs to be `Sync`. If the `Poll` type is split into two types, then `Poll::poll` can take `&mut self` and the lock can be removed.

** But, what about polling from multiple threads? **

Well, that isn't supported today anyway. I also suspect that the answer isn't to make `Poll` `Sync` but to add some clone type functionality so that each thread can get its own `Poll` handle. This would also better support the max concurrency settings that IOCP has.

**Comment by cramertj** [comment]
FWIW, Fuchsia ports are designed for polling from multiple threads, and it would be nice to expose that functionality somehow.

**Comment by carllerche** [comment]
Well, it isn’t really supported today :-) exposing concurrent polling in a portable way is tricky and out of scope. Individual platforms could potentially offer extensions to provide the capability in the mean time.

**Comment by cramertj** [comment]
Yes, I'm just commenting because of your note about offering some kind of cloneable-poll handle. Depending on the design, I could potentially make the handles contain an Arc<Poll> which could allow the user to poll from multiple threads.

**Comment by alexcrichton** [comment]
I thought the extra synchronization logic in `Poll` was just a stopgap? All platforms provide the ability for multiple threads (afaik at least) to get blocked at the system layer, so it seems like something that should be exposed by mio?

**Comment by carllerche** [comment]
@alexcrichton Perhaps, but I don't think it makes sense to enable `Sync` access to poll, especially given that windows has a max concurrency setting and early versions of epoll don't do the right thing. A better option will be to add a "clone" like functionality to `Poll` to be able to hook into the duplication of the handle to guard against things like max concurrency.

**Comment by stjepang** [comment]
@carllerche What you're saying makes sense me.

If one really wants multiple threads to be able to "drive" the same `Poll` instance, perhaps they shouldn't do that in parallel. Instead, have a shared `Arc<Mutex<Poll>>` and then whatever thread has slack time would try locking the `Poll` and driving it. If it's already locked by another thread, then go to sleep.

Another strategy is to have multiple `Poll`s -- one per thread and each thread is driving its own `Poll` when it has slack time, which is exactly what we're doing right now in the default runtime (`tokio::runtime::Runtime`).

Is my thinking right here? Is it ever more desirable to really drive the same `Poll` at the same time from multiple threads?

**Comment by cramertj** [comment]
@stjepang I'm not sure about other platforms, but on Fuchsia it is certainly more performant for multiple threads to share a single zircon Port object and all wait, allowing the kernel to choose which one to wake with which messages. Events can then be delivered to any thread.

**Comment by carllerche** [comment]
@cramertj the eventual plan is to permit concurrent polling by cloning the `Poll` handle (though, w/ some API like `try_clone(&self) -> Option<Self>`. If the platform supports concurrent polling, it can implement clone. This will also allow platforms that support concurrent polling up to a maximum number of threads to enforce this.
